### ISSUE TITLE
Consider changing component Render() sig to return ComponentOrHTML?

### ISSUE THREAD

**Issue by pdf** [gopherjs/vecty#148]
I was looking at a declarative router implementation, and if we defined the return of `Component.Render()` as `vecty.ComponentOrHTML` instead of `*vecty.HTML`, it would enable the writing of 'wrapper', or 'higher-order' components, that enhance the components they wrap with additional functionality, without requiring unwanted DOM elements.  This could come in handy for component libraries, event handling wrappers, routers, etc, whilst allowing the wrapped component to function fully with lifecycle events, persistence, etc.

**Comment by slimsag** [comment]
I haven't considered this very much, and I believe what you are saying to be true, but it would be very helpful to me if you could provide one such code example demonstrating how this would be useful?

**Comment by slimsag** [comment]
@pdf I've taken a look at all open issues and this one sticks out as the last remaining one that would have widespread API breakage.

I've thought about this issue further and the motivation makes more sense to me now. I think this would be a good thing to do, but I can imagine the implementation will prove to be a bit tricky.

**Comment by pdf** [comment]
I believe you're right on all counts.

I wrote this issue when I was short on time, so I should mention for completeness that there are ways to implement some of the use-cases that this would solve, for example proxying lifecycle events and embedding a component in the wrapper.  But I raised this issue after playing with them, because it's not obvious, breaks if the embedded component is nil, etc:

```go
type MyWrapper struct {
  vecty.Component
}

func (w *MyWrapper) Mount() {
  doLocalStuff()
  if m, ok := w.Component.(vecty.Mounter); ok {
    w.Component.Mount()
  }
}
```

I had a pretty good idea at high level of how this change might fit together, and that seems to have born out - I whipped up a change for it in the last hour or so, but I'm probably not going to have time to finish it today.

It's near fully functional - I believe the only remaining tasks are dealing with nested unmounts, and adding tests. I'll push if I get some time tonight to finish it, otherwise it'll probably have to wait until next week.
